### PR TITLE
[feat] cuda graph support and refactor non-functional api

### PR DESCRIPTION
**Pull request by LRY89757**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

1. Add CUDA Graph Support
2. Refactor some non-functional API to make cuda graph could capture the graph

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by LRY89757** [comment]
Please don't merge, still a little bugs to solve. But feel free to review because I've changed some api :)

File: colossalai/kernel/triton/rms_layernorm.py [hunk 1]
```diff
@@ -94,7 +92,10 @@ def _rmsnorm_with_residual_kernel(
 
     def rms_layernorm(x, weight, eps, norm_output=None, residual=None):
         # allocate output
-        y = torch.empty_like(x) if norm_output is None else norm_output
+        # y = torch.empty_like(x) if norm_output is None else norm_output
```
Comment (FrankLeeeee) [inline colossalai/kernel/triton/rms_layernorm.py:95]: Remove this line?

Comment (LRY89757) [reply colossalai/kernel/triton/rms_layernorm.py:95]: ok

**Comment by LRY89757** [comment]
All bugs are fixed :)

**Comment by LRY89757** [comment]
Fix the bugs of dynamic grid for flash decoding, now it passed all the tests and could be merged :).

**Comment by LRY89757** [comment]
Unit Test:

image

**Comment by LRY89757** [comment]
All Unit Test:

image

File: tests/test_infer/test_cuda_graph.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            use_cuda_graph=False,
+            block_size=16,
+        )
+
+    inference_engine = InferenceEngine(model, tokenizer, inference_config, verbose=True)
+    assert inference_engine.generation_config.max_new_tokens == output_len
+    generation_config = GenerationConfig(do_sample=do_sample, top_p=top_p, top_k=top_k)
+    outputs = inference_engine.generate(prompts_token_ids=prompts_token_ids, generation_config=generation_config)
+
+    # print(f"outputs, use_cuda_grpah is {use_cuda_graph}, output: {outputs}")
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_cuda_graph.py:71]: remove print

File: colossalai/inference/core/engine.py [hunk 3]
```diff
@@ -81,11 +86,91 @@ def __init__(
...
+        if self.use_cuda_graph:
+            self.graph_runners: Dict[int, CUDAGraphRunner] = {}
+            self.graph_memory_pool = None  # Set during graph capture.
+            if verbose:
+                self.logger.info("Colossal AI CUDA Graph Capture on")
+
+            self.capture_model(self.k_cache, self.v_cache)
+
+    @torch.inference_mode()
+    def capture_model(self, k_cache: torch.Tensor, v_cache: torch.Tensor):
```
Comment (yuanheng-zhao) [inline colossalai/inference/core/engine.py:104]: The types of `k_cache` and `v_cache` are actually `List[torch.Tensor]` if you use `self.request_handler.get_kvcache()`

File: colossalai/inference/core/engine.py [hunk 4]
```diff
@@ -278,26 +363,76 @@ def add_request(
...
+                (sequence_lengths.sum().item(), batch.num_heads * batch.head_dim),
+                dtype=batch.dtype,
+                device=batch.device,
+            )
+        else:
+            output_tensor = torch.zeros(
+                (batch.current_batch_size, batch.num_heads * batch.head_dim), dtype=batch.dtype, device=batch.device
+            )
+
+        # only when we have the graph for specific decoding batch size can we use the cuda graph for inference
```
Comment (yuanheng-zhao) [inline colossalai/inference/core/engine.py:381]: Might want to add this NOTE into config as well

File: colossalai/inference/core/engine.py [hunk 5]
```diff
@@ -278,26 +363,76 @@ def add_request(
...
+            kv_seq_len=sequence_lengths.max().item(),
+            head_dim=batch.head_dim,
+        )
+
+        # if not batch.is_prompts:
+        #     self.logger.info(f"decoding")
+        #     self.logger.info(f"input metadata is: {input_meta_data}")
+        # else:
+        #     self.logger.info(f"prefill")
+        #     self.logger.info(f"input metadata is: {input_meta_data}")
```
Comment (yuanheng-zhao) [inline colossalai/inference/core/engine.py:403]: Might want to remove testing logs

File: colossalai/inference/core/engine.py [hunk 6]
```diff
@@ -278,26 +363,76 @@ def add_request(
...
             List[str]: Decoded finished sequences generated by one step.
         """
 
         batch = self.request_handler.schedule()
 
+        input_token_ids, output_tensor, input_meta_data = self.prepare_input(batch)
+
+        if input_meta_data.use_cuda_graph:
+            model_executable = self.graph_runners[input_meta_data.batch_size]
+            # self.logger.info("run cuda graph")
```
Comment (yuanheng-zhao) [inline colossalai/inference/core/engine.py:426]: Same as above. Either uncomment the info logs (really logging info for users) or remove those logs (if you think these're only for your testing purpose)

**Comment by LRY89757** [comment]
* After Fix Conflicts with other PR, pass all unit test:
image

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

**Pull request merged**